openadapt/ml_models/openai_module/gpt3_davinci/finetune.py

```diff
@@ -0,0 +1,79 @@
+from openadapt import models
+from openadapt.crud import get_action_events, get_window_events
+import fire
+import openai
+from copy import deepcopy
+import utils
+
+class EventType:
+    WINDOW = "WINDOW"
+    ACTION = "ACTION"
+
+def gpt_davinci_finetune(recording: models.Recording):
+
+   
+    # prompt is ref_win + ref_Act, 
+    # completion is target_win + target_act
+
+    for ref_win, ref_act, target_win, target_act in get_finetune_pairs(recording):
+        pass
+
+    # TODO: create JSON in the format {prompt: ref_win, ref_act completion: target_win, target_act} as per GPT finetune docs.
+
+
+def get_finetune_pairs(recording: models.Recording):
+
+    processed_acx = get_action_events(recording)
+    for prev_action, next_action in zip(processed_acx, processed_acx[1:]):
+        
+        prev_window = prev_action[1]
+        next_window = next_action[1]
+
+        ref_win = prev_window
+        ref_act = prev_action
+
+        target_win = next_window
+        target_act = next_action
+
+        yield ref_win, ref_act, target_win, target_act
+
+
+def sanitize_dicts(reference_action):
+    reference_window = reference_action.window_event
+
+    reference_window_dict = deepcopy({
+        key: val
+        for key, val in utils.row2dict(reference_window, follow=False).items()
+        if val is not None
+        and not key.endswith("timestamp")
+        and not key.endswith("id")
+        #and not isinstance(getattr(models.WindowEvent, key), property)
+    })
+
+    if reference_action.children:
+        reference_action_dicts = [
+            deepcopy({
+                key: val
+                for key, val in utils.row2dict(child, follow=False).items()
+                if val is not None
+                and not key.endswith("timestamp")
+                and not key.endswith("id")
+                and not isinstance(getattr(models.ActionEvent, key), property)
+            })
+            for child in reference_action.children
+        ]
+    else:
+        reference_action_dicts = [
+            deepcopy({
+                key: val
+                for key, val in utils.row2dict(reference_action, follow=False).items()
+                if val is not None
+                and not key.endswith("timestamp")
+                and not key.endswith("id")
+                #and not isinstance(getattr(models.ActionEvent, key), property)
+            })
+        ]
+
+    # TODO: find way to condense data inside state key, as it eats up context length
```

Comment:
@LaPetiteSouris WDYT about the structure of this finetuning pipeline? (Currently implemented only for the GPT 3.5 Davinci model as GPT-4 is not available for finetuning.) 

We finetune autoregressively here but sanitizing the dicts will need some more work as there is a lot of context being eaten up by the meta and data keys of each Window Event.

Reply:
This is exactly my pain point when working on #421 . To formalize the problem:

- It is hard to decide which data to go into the evaluation/tuning pipelines, as we don't know which peace of information may work and which not. Also, each piece of info given will change the output of the models.
- I think there has been efforts on trying to "solve" the model output problems by using guardrails, which was not as we expected. 

I am thinking that we should, for now, limit the scope to the pipeline only, without worrying about the data.

How to sanitize data, which data to put in, to get state, windows or not, is _not_ the scope of this issue. Here we try to define the pipelines and let's not worry much about the performance of the model given the data training.

We can even get simple as just using `action` and ignoring for now `window`. As soon as we get this pipeline, we can experiment things all together and come up with an idea of how to represent the data, how to summarize the data into prompts... etc

In short, let's just use `action`, or strip `windows` to its core for the sake of achieving the fine-tuning pipelines.

Thoughts ? @FFFiend @abrichr

Reply:
PS: To complete, or extend this `sanitize` or `summarize` method for each model, IMHO, is the responsibility of the folk who pick up the subject of fine-tuning a specific model, not you who are putting the pipelines in place.

Reply:
In that case, the pipeline depends currently on the providers and models that we wish to use meaning this is blocked by #419 (correct me if I'm wrong), and extending ```ml_models.py``` itself with more useful high level functions. To that end, I thought of creating an ```add_model``` method to the addition of a provider directory to the user's local ```ml_models``` folder with finetune, infer files along with an addition to the ML_MODELS list with all the necessary info (such as max context length, etc etc). 

Thoughts?

Reply:
I think for now, let's keep `ML_MODELS` as it is for the sake of simplicity and move forward.

The next steps may be is to represent this in a form of `pydantic` class. I think database can be an option in the future, but right now we may benefit from restricting the scope of this issue.

For the models which need new provider, we can go in with 2 steps:

- Add a new PR on top of this one later to extend `ml_models`
- Run #419 with the new models.

IHMO, it is the responsibility of anyone who picks up fine-tuning or RF learning (me in this case 👍  )  to add a PR on top to extend `ml_models` and `sanitize` method to make use of the required models.

Reply:
> In short, let's just use `action`, or strip `windows` to its core for the sake of achieving the fine-tuning pipelines.
> 
> 
> 
> Thoughts ? @FFFiend @abrichr 
> 
> 
@LaPetiteSouris 
I agree, I've done a decent amount of work with the window states and can conclude that there's a lot of unnecessary information for each state -- and some good. There's also OS-specific labels (no Apple Foundation classes on Windows) that may make our work less consistent across platforms. We could strip the window states to, at a minimum:
- window title; ex. opening google chrome and going to outlook would result in a title like Outlook - 10 emails
- width & height of the window: to ensure that our replay strategies are independent of screen sizes
- the "top/left" parts of the window rectangle, and maybe the location of the close button (stored as AXValue)

Reply:
I was thinking of straight up removing this to begin with, and keeping everything else since this isn't especially used inside StatefulReplayStrategy (only key presses and clicks are relevant)

Thoughts? @0dm

Reply:
Let's strip them to bare bone to make an MVP out of it. Also, it helps a lot in keeping the prompt under the 1024 char limit in some cases.

Reply:
We could also..summarize the AXValue key instead of each window state 😄 Related to #430 .